File: di-ipv-cimit-stub/lambdas/put-contra-indicators/src/main/java/uk/gov/di/ipv/core/putcontraindicators/service/ContraIndicatorsService.java
package uk.gov.di.ipv.core.putcontraindicators.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.nimbusds.jwt.SignedJWT;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.message.StringMapMessage;
import uk.gov.di.ipv.core.library.persistence.items.CimitStubItem;
import uk.gov.di.ipv.core.library.service.CimitStubItemService;
import uk.gov.di.ipv.core.library.service.ConfigService;
import uk.gov.di.ipv.core.putcontraindicators.domain.PutContraIndicatorsRequest;
import uk.gov.di.ipv.core.putcontraindicators.dto.CredentialSubject;
import uk.gov.di.ipv.core.putcontraindicators.dto.Evidence;
import uk.gov.di.ipv.core.putcontraindicators.dto.VcClaim;
import uk.gov.di.ipv.core.putcontraindicators.exceptions.CiPutException;

import java.text.ParseException;
import java.time.Instant;
import java.util.Date;
import java.util.List;

import static uk.gov.di.ipv.core.library.vc.VerifiableCredentialConstants.VC_CLAIM;

public class ContraIndicatorsService {
    private static final Logger LOGGER = LogManager.getLogger();
    private static final String LOG_MESSAGE_DESCRIPTION = "description";
    private static final String LOG_ERROR_DESCRIPTION = "errorDescription";
    private static final String LOG_ITEM_COUNT = "itemCount";
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
    private final ConfigService configService;
    private final CimitStubItemService cimitStubItemService;

    public ContraIndicatorsService() {
        this.configService = new ConfigService();
        this.cimitStubItemService = new CimitStubItemService(configService);
    }

    public ContraIndicatorsService(
            ConfigService configService, CimitStubItemService cimitStubItemService) {
        this.configService = configService;
        this.cimitStubItemService = cimitStubItemService;
    }

    public void addUserCis(PutContraIndicatorsRequest request) throws CiPutException {
        try {
            var signedJWT = parseSignedJwt(request);
            var cimitStubItems = toCimitStubItems(signedJWT);
            cimitStubItems.forEach(cimitStubItemService::persistCimitStubItem);
            LOGGER.info(
                    new StringMapMessage()
                            .with(
                                    LOG_MESSAGE_DESCRIPTION,
                                    "Inserted User CI data to the Cimit Stub DynamoDB Table.")
                            .with(LOG_ITEM_COUNT, cimitStubItems.size()));
        } catch (Exception ex) {
            throw new CiPutException(ex.getMessage());
        }
    }

    private SignedJWT parseSignedJwt(PutContraIndicatorsRequest request) throws CiPutException {
        try {
            return SignedJWT.parse(request.getSignedJwt());
        } catch (ParseException e) {
            LOGGER.error(
                    new StringMapMessage()
                            .with(
                                    LOG_ERROR_DESCRIPTION,
                                    "Failed to parse ContraIndicators JWT. Error message:"
                                            + e.getMessage()));
            throw new CiPutException("Failed to parse JWT");
        }
    }

    private VcClaim parseVcClaim(SignedJWT signedJWT) throws CiPutException {
        VcClaim vcClaim;
        try {
            vcClaim =
                    OBJECT_MAPPER.convertValue(
                            signedJWT.getJWTClaimsSet().getClaim(VC_CLAIM), VcClaim.class);
        } catch (ParseException e) {
            LOGGER.info(
                    new StringMapMessage()
                            .with(LOG_MESSAGE_DESCRIPTION, "Failed to parse VC claim")
                            .with(LOG_ERROR_DESCRIPTION, e.getMessage()));
            return null;
        }

        List<Evidence> evidence = vcClaim.evidence();
        if (evidence == null || evidence.size() != 1) {
            String message = "Unexpected evidence count.";
            LOGGER.info(
                    new StringMapMessage()
                            .with(LOG_MESSAGE_DESCRIPTION, message)
                            .with(
                                    LOG_ERROR_DESCRIPTION,
                                    String.format(
                                            "Expected one evidence item, got %d.",
                                            evidence == null ? 0 : evidence.size())));
            return null;
        }

        return vcClaim;
    }

    private List<CimitStubItem> toCimitStubItems(SignedJWT signedJWT) throws ParseException {
        var vcClaim = parseVcClaim(signedJWT);
        if (vcClaim == null) {
            LOGGER.info(
                    new StringMapMessage()
                            .with(LOG_MESSAGE_DESCRIPTION, "No VC claim found in jwt"));
            return List.of();
        }

        var ci = vcClaim.evidence().get(0).ci();
        if (ci == null || ci.isEmpty()) {
            LOGGER.info(new StringMapMessage().with(LOG_MESSAGE_DESCRIPTION, "No CI in VC"));
            return List.of();
        }

        var userId = signedJWT.getJWTClaimsSet().getSubject();
        var iss = signedJWT.getJWTClaimsSet().getIssuer();
        var issuanceDate = getIssuanceDate(signedJWT.getJWTClaimsSet().getNotBeforeTime());
        var docId = getDocumentIdentifier(vcClaim);

        return ci.stream()
                .distinct()
                .map(
                        ciCode ->
                                CimitStubItem.builder()
                                        .userId(userId)
                                        .contraIndicatorCode(ciCode.toUpperCase())
                                        .issuer(iss)
                                        .issuanceDate(issuanceDate)
                                        .mitigations(List.of())
                                        .documentIdentifier(docId)
                                        .build())
                .toList();
    }

    private Instant getIssuanceDate(Date nbf) {
        if (nbf != null) {
            return nbf.toInstant();
        }
        return Instant.now();
    }

    private String getDocumentIdentifier(VcClaim vcClaim) {
        CredentialSubject credentialSubject = vcClaim.credentialSubject();
        if (credentialSubject.drivingPermit() != null
                && !credentialSubject.drivingPermit().isEmpty()) {
            return credentialSubject.drivingPermit().get(0).toIdentifier();
        }
        if (credentialSubject.passport() != null && !credentialSubject.passport().isEmpty()) {
            return credentialSubject.passport().get(0).toIdentifier();
        }
        return null;
    }
}

File: di-ipv-cimit-stub-ts/lambdas/src/common/contraIndicatorsService.ts
import { PutContraIndicatorRequest } from "../internal-api/put-contra-indicators/putContraIndicatorsHandler";
import { getErrorMessage } from "./apiResponseBuilder";
import { decodeJwt } from "jose";
import {
  DrivingPermitDetailsClass,
  IdentityCheckCredentialClass,
  PassportDetailsClass,
  ResidencePermitDetailsClass,
  SocialSecurityRecordDetailsClass,
} from "@govuk-one-login/data-vocab/credentials";
import { createStubItem, persistCimitStubItem } from "./cimitStubItemService";
import { CimitStubItem } from "./contraIndicatorTypes";

export const addUserCIs = async (
  request: PutContraIndicatorRequest,
): Promise<void> => {
  try {
    const signedJwt = decodeJwt(request.signed_jwt);
    const vc = signedJwt.vc as IdentityCheckCredentialClass;
    if (!vc) {
      console.info("No VC claim found in jwt");
      return;
    }
    if (!vc.evidence || vc.evidence.length === 0) {
      console.info("No evidence in VC");
      return;
    }
    const cis = vc.evidence[0].ci;
    if (!cis || cis.length === 0) {
      console.info("No CI in evidence");
      return;
    }

    const issuanceDate = signedJwt.nbf
      ? new Date(signedJwt.nbf * 1000)
      : new Date();
    const issuanceDateString = issuanceDate.toISOString();

    const dbItems: CimitStubItem[] = [];
    for (const ci of cis) {
      const item = await createStubItem(
        signedJwt.sub,
        ci.toUpperCase(),
        signedJwt.iss,
        issuanceDateString,
        [],
        getDocument(vc),
        vc.evidence[0].txn,
      );
      dbItems.push(item);
    }

    for (const item of dbItems) {
      await persistCimitStubItem(item);
    }
  } catch (error) {
    console.error(getErrorMessage(error));
    throw new Error("Failed to add user CIs");
  }
};

const getDocument = (vc: IdentityCheckCredentialClass): string => {
  const credentialSubject = vc.credentialSubject;
  if (!credentialSubject) return "";
  if (
    !!credentialSubject.drivingPermit &&
    credentialSubject.drivingPermit.length > 0
  ) {
    const drivingPermit: DrivingPermitDetailsClass =
      credentialSubject.drivingPermit[0];
    return `drivingPermit/GB/${drivingPermit.issuedBy}/${drivingPermit.personalNumber}/${drivingPermit.issueDate}`;
  }
  if (!!credentialSubject.passport && credentialSubject.passport.length > 0) {
    const passport: PassportDetailsClass = credentialSubject.passport[0];
    return `passport/${passport.icaoIssuerCode}/${passport.documentNumber}`;
  }
  if (
    !!credentialSubject.residencePermit &&
    credentialSubject.residencePermit.length > 0
  ) {
    const residencePermit: ResidencePermitDetailsClass =
      credentialSubject.residencePermit[0];
    return `residencePermit/${residencePermit.icaoIssuerCode}/${residencePermit.documentNumber}`;
  }
  if (
    !!credentialSubject.socialSecurityRecord &&
    credentialSubject.socialSecurityRecord.length > 0
  ) {
    const socialSecurityRecord: SocialSecurityRecordDetailsClass =
      credentialSubject.socialSecurityRecord[0];
    return `socialSecurity/GB/${socialSecurityRecord.personalNumber}`;
  }
  return "";
};
